File: find_max_subarray_sum.cpp
#include<iostream>
#include "global_types.h"
#include <limits.h>


using namespace std;


int nums[]= {1, 34 , -10, -40, -2, 10, 9, 10, -7, 10, -5} ;

struct subArr{
    int low;
    int high;
    long sum ; 
} ;

void setVal(subArr &arr, int low, int high, long sum){

}

subArr findMaxCorssSubArr(int * nums, int l, int h){
    subArr MaxSub;
    MaxSub.low=0;
    MaxSub.high=0;
    MaxSub.sum=0;
    return MaxSub;
}


subArr findMaxSubArr(int * nums, int l, int h,subArr &MaxSub){
    subArr MaxSub;
    if (h==l){
        setVal(MaxSub,l,l,nums[l]);
        return MaxSub;
    }
}



int main(){
    int ln= (nums[0]==NULL ? 0 : sizeof(nums)/sizeof(nums[0]) );
    subArr MaxSub;
    findMaxSubArr(nums,0,ln,MaxSub);

    return 0;
}

File: find_max_subarray_sum.py
nums = [1, 34 , -10, -40, -2, 10, 9, 10, -7, 10, -5, 100, 20, -10, 0, -150, -1, 10, -12, 12, 19]

def findMaxCrossSubArr(arr, low, mid, high) :
    
    maxLeft=mid
    maxRight=mid
    
    leftSum = float('-inf')
    sum=0
    for i in range(mid, low-1, -1 ) : 
        sum = sum + arr[i]
        if sum > leftSum : 
            maxLeft=i
            leftSum=sum

    rightSum = float('-inf')
    sum=0
    for i in range(mid+1, high+1, 1 ) : 
        sum = sum + arr[i]
        if sum > rightSum : 
            maxRight=i
            rightSum=sum

    return (maxLeft,maxRight,leftSum+rightSum) 

def findMaxSubArr(arr, low, high) :
    
    if low == high : 
        return (low,low,arr[low])
    else :
        mid = ( low + high ) // 2
        leftLow, leftHigh, leftSum = findMaxSubArr(arr, low, mid)
        rightLow, rightHigh, rightSum = findMaxSubArr(arr, mid+1, high)
        crossLow, crossHigh, crossSum = findMaxCrossSubArr(arr, low, mid, high)

        if leftSum >= rightSum and leftSum >=crossSum :
            return (leftLow, leftHigh, leftSum)
        elif rightSum > leftSum and rightSum > crossSum : 
            return (rightLow, rightHigh, rightSum)
        else : 
            return (crossLow, crossHigh, crossSum)


def main():
    print("In main\nINPUT  : {} \nOUTPUT : {}".format(nums, findMaxSubArr(nums,0,len(nums)-1) ))
    return 0

if __name__ == "__main__":
    main()